recipes-app/iot2050-firmware-update/files/iot2050-firmware-update

```diff
@@ -475,27 +549,39 @@ def main(argv):
         erase_env_input = input("\nWarning: All U-Boot environment variables will be reset to factory settings. Continue (y/N)? ")
         if not erase_env_input == "y":
             sys.exit(1)
-
-    if args.force:
-        # When forced, ignore all the checking and call the underlying updater directly
-        updater = FirmwareUpdate(args.firmware)
-    else:
-        updater = ManagedFirmwareUpdate(args.firmware)
-
-    envlist = None
-    if not args.force and not args.reset:
-        envlist = updater.get_preserved_uboot_env(args.preserve_list)
-
-    updater.update_firmware()
-
-    if envlist:
-        print("\n")
-        for env in envlist:
-            print(env)
-        preserved_env_input = input(
-            "\nWarning: Preserve the above env variables. Confirm(Y/n)? ")
-        if not preserved_env_input == "n":
-            updater.update_uboot_env(envlist)
+    try:
+        if args.force:
+            # When forced, ignore all the checking and call the underlying updater directly
+            updater = FirmwareUpdate(args.firmware)
+        else:
+            updater = ManagedFirmwareUpdate(args.firmware)
+
+        envlist = None
+        if not args.force and not args.reset:
+            envlist = updater.get_preserved_uboot_env(args.preserve_list)
+
+        updater.update_firmware()
+
+        if envlist:
+            print("\n")
+            for env in envlist:
+                print(env)
+            preserved_env_input = input(
+                "\nWarning: Preserve the above env variables. Confirm(Y/n)? ")
+            if not preserved_env_input == "n":
+                updater.update_uboot_env(envlist)
+    except UpgradeError as e:
+        print(e)
+        intput_reminder = '''
+        ====================================
+        Upgrade Failed! Please do not reboot!!
+        The Device may become brick.
+        Please try upgrade firmware again!
```

Comment:
How likely is it that retrying will resolve? I would also recommend re-downloading.

Reply:
Flash programming error is not so rare, if the downloading has no error, then most likely the error is programming, such retrying should helps.

Of course, a re-downloading recommendation should also helps. Or we can suggest the user to calculate the checksum after downloading, before upgrading, in the document.

Reply:
Do we have a checksum as part of the managed update package? If so, my comment is indeed unneeded.

Reply:
No, since a broken tar.xz should not be extract-able. Or is it?

However, if we think a checksum is helping then since we have the so-called `managed update`, we can  enhance the `update.conf.json` to have the checksum included. However, I suggest to have it later since the current status is sufficient to detect at least the programming error.